Dropout

anderspapitto opened the pull request:
(most of this diff is the elman rnn that it's based on - the actual diff is just deleting the lockdowns

houseroad commented on torch/onnx/symbolic.py:
```diff
@@ -534,6 +534,8 @@ def RNN_symbolic_builder(cell_type, *args, **kwargs):
         return LSTM_symbolic_builder(*args, **kwargs)
     elif cell_type == 'GRU':
         return GRU_symbolic_builder(*args, **kwargs)
+    elif cell_type.startswith('RNN_'):
```
Here the meaning of "RNN_" should be the in-place version. Probably you should handle both RNN_ and RNN.

anderspapitto commented:
@houseroad that's not part of this diff, that's part of the underlying https://github.com/pytorch/pytorch/pull/4613

This diff Is only the part that unlocks dropout

ezyang commented:
@pytorchbot retest this please

ezyang commented:
Lint failed.

ezyang commented:
Is this legit?  You have a different network if you don't have dropout, and that's effectively what this patch is doing.

Perhaps what you actually wanted to do is ignore dropout in inference mode, because in that case no dropout occurs. That's OK (test for THAT case), and error if someone tries to export in training mode. An example where we do this test already is FeatureDropout.

ezyang commented on torch/onnx/symbolic.py:
```diff
@@ -543,11 +545,43 @@ def reform_weights(g, w, n, intervals):
     return g.op('Concat', *slices, axis_i=0)
 
 
+def Elman_RNN_symbolic_builder(nonlinearity, input_size, hidden_size, num_layers, batch_first, dropout, bidirectional, **kwargs):
```
I understand that `kwargs` shows up in the other RNN implementations, but is there a reason it is done this way? It seems safer to explicitly list the arguments, so that if someone adds a new argument they will notice the symbolic needs to be updated.

Commit:
PackedSequence: store batch_sizes as tensor

rather than converting to a list of python integers. This maintains
the invariant that module's inputs/outputs are collections of
Variables.

In particular, this causes the JIT to no longer choke when flattening
and unflattening arguments.

anderspapitto committed:
Handle sequence lengths correctly when exporting RNNs to ONNX

- when uniform sequence lengths are provided, correctly omit the
  argument when constructing the ONNX graph, so as to not fix the
  graph to the batch size.

- handle PackedSequences by floating them through the graph and
  eliminating them in an optimization pass. ONNX does not have packed
  sequences, but operates on a representation equivalent to
  PaddedSequence, so we hide the representation-switching from ONNX

- as a preliminary step towards handling PackedSequences, not directly
  tied to ONNX export, change batch_sizes from being an argument to
  the RNN operators into being an argument to the forward() function
  of those RNN operators. This more closely models the reality that
  batch_sizes are effectively part of the input sequences.

anderspapitto committed:
enable dropout in export of RNNs to ONNX
